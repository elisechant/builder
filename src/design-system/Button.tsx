import type {HTMLProps, ReactNode} from "react";
import {twMerge} from "tailwind-merge";

const BASE = 'focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2'
const BLOCK = 'w-full justify-center'
const SIZE = {
  SM: 'text-sm',
  LG: 'text-lg',
}
const COLOR = {
  PRIMARY: 'text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800',
  SECONDARY: 'text-white bg-gray-600 hover:bg-gray-700 disabled:bg-gray-600 focus-visible:outline-gray-600 border-b-gray-800',
}

type NativeProps = Omit<HTMLProps<HTMLButtonElement>, 'size' | 'type'> & {
  type?: "submit" | "reset" | "button"
}

export interface ButtonProps extends NativeProps {
  children: ReactNode
  color?: 'primary' | 'secondary'
  size?: 'sm' | 'lg'
  block?: boolean
}

export default function Button({ children, color = 'primary', size = 'sm', block = false, ...props }: ButtonProps) {
  let tw = [BASE]
  if (block) {
    tw.push(BLOCK)
  }

  if (size === 'sm') {
    tw.push(SIZE.SM)
  } else if (size === 'lg') {
    tw.push(SIZE.LG)
  } else {
    throw new Error(`size:${size} not handled`)
  }
  if (color === 'primary') {
    tw.push(COLOR.PRIMARY)
  } else if (color === 'secondary') {
    tw.push(COLOR.SECONDARY)
  } else {
    throw new Error(`color:${color} not handled`)
  }

  return (
    <button {...props} className={twMerge(tw)}>{children}</button>
  )
}

