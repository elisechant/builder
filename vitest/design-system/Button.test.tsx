import {describe, expect, test} from 'vitest'
import Button from '@/design-system/Button'
import {render} from '@testing-library/react'

const renderButton = (props = {}) => render(<Button {...props}>text</Button>)

describe('Button', () => {
  describe('style variants', function () {
    test('happy path', () => {
      expect(renderButton()).toMatchSnapshot()
    })
    test('block', () => {
      expect(renderButton({ block: true })).toMatchSnapshot()
    })
    test('color', function() {
      expect(renderButton({ color: 'primary' })).toMatchSnapshot()
      expect(renderButton({ color: 'secondary' })).toMatchSnapshot()
      expect(() => renderButton({ color: 'tertiary' })).toThrowError()
    })
    test('size', function() {
      expect(renderButton({ size: 'sm' })).toMatchSnapshot()
      expect(renderButton({ size: 'lg' })).toMatchSnapshot()
      expect(() => renderButton({ size: 'md' })).toThrowError()
    })
  });
})
