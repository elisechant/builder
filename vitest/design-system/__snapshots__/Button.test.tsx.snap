// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > style variants > block 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 w-full justify-center text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 w-full justify-center text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button > style variants > color 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button > style variants > color 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-gray-600 hover:bg-gray-700 disabled:bg-gray-600 focus-visible:outline-gray-600 border-b-gray-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-gray-600 hover:bg-gray-700 disabled:bg-gray-600 focus-visible:outline-gray-600 border-b-gray-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button > style variants > happy path 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button > style variants > size 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button > style variants > size 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-sm text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
    <div>
      <button
        class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-lg text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
      >
        text
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-normal rounded-sm gap-x-1.5 px-2.5 py-2.5 inline-flex items-center focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 border-b-2 text-lg text-white bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 focus-visible:outline-blue-500 border-b-blue-800"
    >
      text
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
